Subject: tty deadlock in kernel-xen, xenconsole
References: bnc#726333
From: <ohering@suse.de>
Patch-mainline: never

xenconsole is broken. tty_mutex does not nest into big tty mutex.
This is violated in tty_release->xencons_close.

---
 drivers/xen/console/console.c |    7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

--- sle11sp2-2011-11-04.orig/drivers/xen/console/console.c	2011-02-02 15:10:34.000000000 +0100
+++ sle11sp2-2011-11-04/drivers/xen/console/console.c	2011-11-04 15:04:09.000000000 +0100
@@ -618,16 +618,11 @@ static void xencons_close(struct tty_str
 	if (DUMMY_TTY(tty))
 		return;
 
-	mutex_lock(&tty_mutex);
-
-	if (tty->count != 1) {
-		mutex_unlock(&tty_mutex);
+	if (tty->count != 1)
 		return;
-	}
 
 	/* Prevent other threads from re-opening this tty. */
 	set_bit(TTY_CLOSING, &tty->flags);
-	mutex_unlock(&tty_mutex);
 
 	tty->closing = 1;
 	tty_wait_until_sent(tty, 0);
